proto/docs/p4runtime.md

```diff
@@ -515,3 +515,164 @@ messages refer to singleton ports by default.
+targets, a centralized SDN controller may define its own types and numbering of
+such PSA intrinsic metadata. In particular, different PSA devices may have
+different types and numbering schemes for ports and class of service. For such
+metadata, a translation between the controller's metadata values and the
+corresponding target specific metadata values is required at runtime.
+
+![Controllers with P4Runtime channels to multiple PSA devices](sdn_controller.svg)
+
+The above figure illustrates a motivating example, where an SDN controller is
+controlling multiple P4 targets in a fabric. Switches 1, 2, 3 etc.
+are all targets with different PSA devices, each defining its own port and
+class of service types. In order to uniformly program the fabric, the
+SDN controller may define 32 bit type for ports and 8 bit type for class of
+service. The controller may also define a common number space for ports and
+cannonical classes of service across the fabric. The PSA devices may define
+much shorter bitwidths for these metadata in order to save precious on-chip
+memory in the switch silicon. Furthermore, the values of ports and classes of
+service in the dataplane are PSA device specific and may be different from the
+values used in the controller(s).
+
+In order to support such SDN use case, P4Runtime requires translation of port
+and class of service metadata values between the controller's space and the
+PSA device's space as needed. Such translation is enabled by a standard
+P4Runtime annotation `@p4runtime_translation`. The annotation value is a
+string identifying the metadata type to be translated. Specifically, we define
+* `@p4runtime_translation("port")` for port translation.
+* `@p4runtime_translation("cos")` for class of service translation.
+
+The actual translation logic resides in the configuration of the PSA target.
+The P4Runtime service on the target is expected to use the mapping defined
+in the configuration to perform runtime translation between the metadata
+values in the control plane and the dataplane for P4 objects that have the
+above annotations defined.
+
+The sub-sections below provide comprehensive examples for various translation
+scenarios.
+
+### Translation of Header Fields
+Port and class of service type fields can be part of header types. For example,
+a packet-out header from the controller to the switch may define the egress
+port of the packet as shown below.
+```
+@controller_header("packet_out")
+header PacketOut_t {
+  @p4runtime_translation("port")
+  PortIdInHeader_t egress_port;
+}
+```
+The top annotation `@controller_header` is also a standard P4Runtime annotation
+that identifies a header type for a controller packet-out. When the target
+receives a packet-out from the controller over the P4Runtime stream channel, it
+will normally expect packet-out metadata (`egress_port`) value of bitwidth as
+defined by the PSA target's `PortIdInHeader_t` type and in the space of the
```

Comment:
I thought part of the rationale for having the `InHeader_t` types was that it enabled using the type / bitwidth that the controller actually wanted?
So for this specific case, there will be no type translation, just a value translation, no? (the type "translation" will be a cast in the P4 program, not a software conversion in the agent?)

Reply:
My only reason for proposing the InHeader_t types is that they should be a common bit width for _all_ PSA devices.  Yes, if we make them too narrow, then that is bad, which is why I keep asking people to double-check the bit widths chosen at the moment so they are not too small for future uses.

If they are wide enough for all PSA devices, then those widths _could_ also be used as the bitwidths of these fields when present in packet headers for PacketIn / PacketOut packets, and be a common width across all controllers, too.  It isn't necessary, just a possibility.

Reply:
+1 to @jafingerhut. It is ok for `PortInHeader_t` to be 16 bits and the controller's port type to be 32-bits.